src/algorithms/libs/beidou_b1c_signal_processing.cc

```diff
@@ -0,0 +1,913 @@
+	
+	int32_t sinboc_11[12 * BEIDOU_B1Cp_CODE_LENGTH_CHIPS] = {0};     //  _codeLength not accepted by Clang
+    	gsl::span<int32_t> sinboc_11_(sinboc_11, _codeLength);
+    	
+    	// 2. Apply Sine BOC(1,1) on generated Beidou B1C Pilot Code
+    	beidou_b1c_pilot_sinboc_11_gen_int(sinboc_11_, gsl::span<int>(b1c_pilot_primary_code_chips, static_cast<uint32_t>(BEIDOU_B1Cp_CODE_LENGTH_CHIPS)));  //generate sinboc(1,1) 12 samples per chip
+    	
+	// 3. Multiply the output of Sine BOC(1,1) with the Power Ratio
+	for (uint32_t i = 0; i < _codeLength; i++)
+        {
+        	_dest[i] = beta * static_cast<float>(sinboc_11[i]);
+                               
+	}
+}
+
+
+//! Generates complex version of both pilot components having sine BOC(6,1) which is in Real part and sine BOC(1,1) which is in Imaginary part
+void beidou_b1cp_code_gen_complex_sampled_boc_61_11(gsl::span<std::complex<float>> _dest,uint32_t _prn, int32_t _fs)
+{
+    //uint32_t _code_length =12 * BEIDOU_B1Cd_CODE_LENGTH_CHIPS;
+    //int32_t _code[_code_length];
+    
+    int32_t _samplesPerCode, _codeValueIndex;
+    float _ts;
+    float _tc;
+    const int32_t _codeLength =12 * BEIDOU_B1Cp_CODE_LENGTH_CHIPS;
+
+    //--- Find number of samples per spreading code ----------------------------
+    _samplesPerCode = static_cast<int>(static_cast<double>(_fs) / (static_cast<double>(BEIDOU_B1Cp_CODE_RATE_HZ) / static_cast<double>(_codeLength)));
+    
+    auto* real_code = static_cast<float*>(volk_gnsssdr_malloc(_samplesPerCode * sizeof(float), volk_gnsssdr_get_alignment()));
+    gsl::span<float> real_code_span(real_code, _samplesPerCode);
+    beidou_b1cp_gen_float_61(real_code_span, _prn);
+    
+    
+    auto* imaginary_code = static_cast<float*>(volk_gnsssdr_malloc(_samplesPerCode * sizeof(float), volk_gnsssdr_get_alignment()));
+    gsl::span<float> imaginary_code_span(imaginary_code, _samplesPerCode);
+    beidou_b1cp_gen_float_11(imaginary_code_span, _prn);
+    
+     //--- Find time constants --------------------------------------------------
+    _ts = 1.0 / static_cast<float>(_fs);                       // Sampling period in sec
+    _tc = 1.0 / static_cast<float>(BEIDOU_B1Cp_CODE_RATE_HZ);  // code chip period in sec
+    
+        
+    for (uint32_t i = 0; i < _samplesPerCode; i++)
+        {
+            //=== Digitizing =======================================================
+
+            //--- Make index array to read B1C code values -------------------------
+            _codeValueIndex = ceil((_ts * (static_cast<float>(i) + 1)) / _tc) - 1;
+
+            //--- Make the digitized version of the B1Cd code -----------------------
+            if (i == _samplesPerCode - 1)
+                {
+                    //--- Correct the last index (due to number rounding issues) -----------
+                    _dest[i] = std::complex<float>(1.0 - 2.0 * real_code_span[_codeLength - 1], 1.0 - 2.0 * imaginary_code_span[_codeLength - 1]);
```

Comment:
You are using the index _codeLength here, which translates to always using the same index. This seems to be a typo, consider fixing